### PR TITLE
Have `jj log` hightlight unique prefixes, `BTree`-based implementation

### PR DESCRIPTION
**Pull request by ilyagr**
This makes it possible to set `ui.unique-prefixes="underscore"` and make `jj` highlight the unique prefix of each ID as follows:

screenshot

There are some more thoughts in the commit descriptions.

### UI

We should make use of https://github.com/martinvonz/jj/pull/962 and make it possible to mark the prefix with underline & color. The current underscore highlight is still useful for tests, at least. I'm not sure how urgent this is.

In fact, I have a **question**: I cautiously kept the default at no highlighting, mainly since the underscores mess up copying and pasting. Was this over-cautious?

### Use this index in other areas

The IdIndex created in this PR could immediately be used to find divergent commits and to look up commits by prefix replacing these functions.

https://github.com/martinvonz/jj/blob/50321f851f4ce0f484fcd029b4d04581ad356d64/lib/src/revset.rs#L104-L147

I think we should do it, mainly to make it much harder for these parts of `jj` to disagree about what prefix resolves to what change/commit. Let me know if you have concerns.

### Combine IdIndex with the normal Index

They overlap in  functionality. Implementing this properly should be able to speed up the loading of the IdIndex immensely. However, this is a big change that intimidates me a little.

This could use `fst` or https://sapling-scm.com/docs/internals/indexedlog/ (which I just discovered from @arxanas 's writing. It seems to be designed specifically for this purpose, but also seems to be GPL).

### Performance

- Performance issues should only be noticeable on large repos (I tried https://github.com/torvalds/linux). This BTree-based index should be fast enough for repositories the size of Git. Even the naive implementation is fast enough for `jj`.

- There is a draft trie implementation of IdIndex in https://github.com/martinvonz/jj/pull/1041. It's a little more complicated, but 5x faster when measuring the speed of computing prefix for each id. It's similar to this implementation by other measures of performance.

- I have a benchmark to compare different implementations of `IdIndex` that I might add to this PR or submit separately.

# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [x] I have updated the documentation (README.md, docs/, demos/)
- [x] I have added tests to cover my changes

### CODE REVIEW

File: lib/src/settings.rs [hunk 1]
```diff
@@ -143,6 +143,12 @@ impl UserSettings {
             .unwrap_or(false)
     }
 
+    pub fn unique_prefixes(&self) -> String {
```
Comment (ilyagr) [inline lib/src/settings.rs:146]: I am planning to make this return an enum rather than a string, similarly to colors. Haven't quite done it yet.

**Comment by arxanas** [comment]
> This could use fst or https://sapling-scm.com/docs/internals/indexedlog/ (which I just discovered from @arxanas 's writing. It seems to be designed specifically for this purpose, but also seems to be GPL).

The library itself is MIT-licensed now!

What does the index store normally? Perhaps we should replace it entirely with the segmented changelog?

**Comment by ilyagr** [comment]
> The library itself is MIT-licensed now!
>

Nice! What's the right link for it?

> What does the index store normally? Perhaps we should replace it entirely
> with the segmented changelog?
>
From what I can tell, this seems like a reasonable thing to try to do, but
I'm not entirely sure.

**Comment by arxanas** [comment]
> what's the right link for it?

I think it's this one: https://github.com/facebook/sapling/blob/d460daa630f005ef550cf3c04cc649f6e301a16d/eden/scm/lib/dag/Cargo.toml#L7

It's not published to crates.io, so we'd have to take a source dependency on it or republish it ourselves. I'm happy to do the republishing, if necessary.

**Comment by ilyagr** [comment]
I think that's the graph-drawing one. The indexlog one seems to be this one:

https://github.com/facebook/sapling/blob/d460daa630f005ef550cf3c04cc649f6e301a16d/eden/scm/lib/indexedlog/LICENSE

Also MIT.

File: lib/src/repo.rs [hunk 2]
```diff
@@ -241,6 +242,40 @@ impl ReadonlyRepo {
         })
     }
 
+    // An interface for testing this functionality directly is constructed in
+    // a follow-up commit.
+    pub fn shortest_unique_prefix_length(&self, target_id_hex: &str) -> usize {
```
Comment (martinvonz) [inline lib/src/repo.rs:269]: Can we put this on the index instead? Then it will work for `MutableRepo`s too. It may not usually matter, but I also don't think it should be hard to do. Putting it on the index also means that we can take advantage of the lookup table it already has. I think this is the sort of thing the index should be responsible for, so different index implementations can do it differently. In particular, we'll clearly not want to iterate over all commits in Google's internal repo :)

File: CHANGELOG.md [hunk 3]
```diff
@@ -52,6 +52,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 * The `empty` condition in templates is true when the commit makes no change to
   the three compared to its parents.
 
+* Setting the new `ui.unique-prefixes` option to `underscore` makes `jj log`
+  highlight the shortest unique prefix of every commit and change id.
```
Comment (martinvonz) [inline CHANGELOG.md:56]: > For now, `none` is the default, since the `underscore` value messes up copy and pasting of ids.

How about making it a colon for now then? Both iTerm2 and Gnome Terminal stop double-click selection at colon.

File: lib/src/repo.rs [hunk 4]
```diff
@@ -242,38 +245,30 @@ impl ReadonlyRepo {
         })
     }
 
-    // An interface for testing this functionality directly is constructed in
-    // a follow-up commit.
+    fn commit_change_id_index(&self) -> &IdIndex {
```
Comment (martinvonz) [inline lib/src/repo.rs:248]: Since the BTree version is both simpler and better than the naive version, I'd just squash it into that version. But I would still prefer that it's abstracted in the Index type.

By the way, we'll also want a higher-level function (outside the index) for first attempting to resolve a prefix within an configured revset (perhaps `ui.default-revset`). I don't think that will impact the interface we need from the index, however - we'll probably just want exactly the signature of your current `shortest_unique_prefix_length()`.

File: CHANGELOG.md (hunk 3 above)
Comment (ilyagr) [reply CHANGELOG.md:56]: Colon is a good idea -- it does seem to separate words and looks pretty good. Do you think the default should be "none" or "colon"?

Comment (martinvonz) [reply CHANGELOG.md:56]: My general feeling is that we're still early enough in the project that we should experiment quite freely before it's too late :) So my vote is for "colon".

File: lib/src/repo.rs (hunk 4 above)
Comment (ilyagr) [reply lib/src/repo.rs:248]: > By the way, we'll also want a higher-level function (outside the index) for first attempting
> to resolve a prefix within an configured revset (perhaps ui.default-revset)

It's important that whatever does this is used both for finding the shortest possible prefix of a commit and for looking up commit by its prefix.

I'm not sure if revsets are the best for this -- they don't currently cover invisible commits. I was thinking along the lines of restricting to all of your commits for the past year and all of their non-public ancestors/descendants, and all entries in the obslog for these. But all of this can be decided later.

I originally thought we'd have a sequence of larger and larger tries to do this. I think you have a different idea in mind.

Comment (martinvonz) [reply lib/src/repo.rs:248]: > It's important that whatever does this is used both for finding the shortest possible prefix of a commit and for looking up commit by its prefix.

True. The lookup would probably have to be somewhere around here: https://github.com/martinvonz/jj/blob/24ccd80a5c0b3c4ad7306b7db35adf28a2747a91/lib/src/revset.rs#L193-L203

> I'm not sure whether revsets are the best for this -- they don't currently cover invisible commits, and I'd like to have easy access to at least recent entries in my obslog (though I wouldn't say it's critical, if this desire makes life harder).

We could add revsets that access the operation log in order to find those. At Google, our default revset is small enough that the unique prefix is one or two hex digits for most users' workflows.

File: CHANGELOG.md (hunk 3 above)
Comment (avamsi) [reply CHANGELOG.md:56]: +1 for colon instead of none as the default (it's a nice feature that should be on by default IMO), re: copy paste, I have to use PuTTY for work and even that stops at colon, so that's good -- one way to side step this would be of course to use color or something like that for differentiation (grey out the non unique part of the id maybe).

Comment (yuja) [reply CHANGELOG.md:56]: Just fyi. I've seen some people get confused with `49904:f3e95e5a5895` (`{rev}:{node|short}`) in Mercurial because it's also valid as a range expression.

Comment (martinvonz) [reply CHANGELOG.md:56]: Oh, that's a good point! iTerm2 seems to default to "/-+\~_." as part of a word (in addition to letters, presumably). Gnome terminal uses a larger set, but `$`, `^`, and `*` all seem to work in both of those two. I don't really have a preference among them. Anyone else?

Comment (yuja) [reply CHANGELOG.md:56]: I would pick `$` since it's unlikely we would add infix `$` op (and it looks like a mangled symbol), but it might be hard to distinguish from `8`.

FWIW, foot uses a different set of characters. ,│`|:"'()[]{}<> and whitespaces are "delimiters".
https://codeberg.org/dnkl/foot/src/branch/master/doc/foot.ini.5.scd#L306

Comment (ilyagr) [reply CHANGELOG.md:56]: How about `,`? `^` looks weird to me, but not terrible.

I never double-click to select, so I'm not sure how important that is. For selecting the usual way, all symbols are equally not great. Once we get to `$`, I'd personally prefer `_`. `$` looks a bit distracting to me. 

If people prefer it, i can live with it, especially if it's a default that can be changed.  I could implement two or three possible symbols (`$`,`:`,`_`), or make it possible to use any symbol (this might be overkill).

Finally, we could implement underline + bold highlighting, perhaps pretty soon. It doesn't interfere with copy/paste at all.

Comment (martinvonz) [reply CHANGELOG.md:56]: I don't care much about which one we pick. I don't think it's worth making it configurable since it's supposed to just be a temporary thing.

An alternative to underling and/or bold, I like @avamsi's proposal of graying out the remainder of the string. I'm not sure I'll like it better, but let's keep it in mind.

Comment (ilyagr) [reply CHANGELOG.md:56]: I guess , will make it look like there are several ids. I'll add `.` for consideration instead.

Greying out sounds nice, we should certainly consider it.

File: lib/src/repo.rs (hunk 2 above)
Comment (ilyagr) [reply lib/src/repo.rs:269]: I certainly agree it belongs to the index. Is there a way which change ids are visible from inside the index? I would need something to replace

```rust
       let all_visible_revisions = RevsetExpression::all()
            .evaluate(self.as_repo_ref(), None)
            .unwrap();
        let change_hex_iter = all_visible_revisions
            .iter()
            .map(|index_entry| index_entry.change_id().hex());
```

I don't know if there's an easy way to give an `IndexEntry` the ability to know whether it's visible. 

We could also consider changing the behavior of ignoring invisible change ids, but that's a larger change, and I'm not sure if it's for the better.

File: CHANGELOG.md (hunk 3 above)
Comment (ilyagr) [reply CHANGELOG.md:56]: OK, currently `.` is my favorite. 

image


One more option is brackets. I'm not sure which is better:

Bracket A
image

Bracket B
image


If I don't hear otherwise, or think of some problem with it myself, I'm going with `.` for now.

File: lib/src/settings.rs (hunk 1 above)
Comment (ilyagr) [reply lib/src/settings.rs:146]: Actually, I am leaning towards leaving it as is. I tried making it a type, and it didn't look like an improvement since it can't really be part of the `lib` crate and there's only a single place it's used.

File: CHANGELOG.md (hunk 3 above)
Comment (martinvonz) [reply CHANGELOG.md:63]: Bracket B looks good to me. It matches the common expectation that the part within brackets is optional.

File: lib/src/repo.rs (hunk 2 above)
Comment (martinvonz) [reply lib/src/repo.rs:269]: Ah, I forgot about change IDs. I don't remember what we said last time, but do you think it makes sense to have one method on the index for finding a unique prefix among all indexed commit IDs, and another method for a unique prefix among all change IDs reachable from a given set of heads? Then the caller can pass in the visible heads to that second method. I feel like it's better to keep the searches among commit IDs and change IDs separate, so the caller can decide. For example, we might add `commit_id(ab12)` and `change_id(ab12)` revsets one day, and they shouldn't have to resolve the prefixes among the other kind of ID. What do you think?

Comment (ilyagr) [reply lib/src/repo.rs:269]: From the algorithmic point of view, the important thing is that the change ids should be in an index, and in the end state it should be efficiently stored on disk. Currently, all the trie/BTree stuff is mainly needed because change ids aren't stored in an optimized (sorted, at least) fashion on disk.

It would make sense the repo to only store an in-memory index of change ids and use the other index for commit ids. I didn't do it because it makes the logic slightly harder than when everything is in the same bucket, and there's nothing in `jj` that would use the separation just yet.

This is not strictly relevant to this PR, but I think I was also leaning towards putting everything in the same bucket after you told me that the lookup by prefix in the Index is inefficient. It was simpler for me to plan to avoid it than to figure out how to fix it. It would just take adding a few lines to `IdIndex` using either a BTree or a trie.

From the UI point of view, I was thinking of the "IdIndex" I was building as more of a precursor to  what we were talking about https://github.com/martinvonz/jj/pull/994#discussion_r1070874338: a unified place responsible for the logic of how user-specified prefixes correspond to the thing they want to find (usually a change or an id). Such a thing should exist, whether it stores the ids itself or communicates with two different indices is less important. I'm itching to make the same entity responsible for the shortest prefix computation (not the algorithmic part, but the part that controls the logic) and for looking up a commit/change by prefix.

In other words, while eventually different entities should probably be responsible for the algorithmic part and the UI part, for now I tried to make one thing that will do both.

Comment (martinvonz) [reply lib/src/repo.rs:269]: That's fair. We're going to have to rewrite the index soon and then we'll revisit this. Until then, it sounds good to me to just do the simplest reasonable thing.

**Comment by martinvonz** [comment]
Regarding IndexedLog, keep in mind that our storage can currently be safely stored in a distributed file system. It's safe to merge two diverged repos by simply having one overwrite the other, as long as directories are merged in the trivial way (including both sets of files and subdirectories, not ignoring entries from one side). I think we'd lose that property if we started using IndexedLog. We can still consider it, but we should be aware of the tradeoff.

File: lib/src/repo.rs (hunk 4 above)
Comment (ilyagr) [reply lib/src/repo.rs:248]: > Since the BTree version is both simpler and better than the naive version, I'd just squash it into that version

I don't know if that's worth doing at this point. Let me know if you'd really like me to.

File: CHANGELOG.md (hunk 3 above)
Comment (ilyagr) [reply CHANGELOG.md]: Bracket B it is then. I also made it the default since that seemed to be your preference, but I don't feel strongly about that one way or the other.

File: lib/src/repo.rs (hunk 4 above)
Comment (martinvonz) [reply lib/src/repo.rs:248]: Nah, that's fine. I've already reviewed both versions now, but it would have been better to not have to review the naive version :)

File: CHANGELOG.md (hunk 3 above)
Comment (martinvonz) [reply CHANGELOG.md]: I also don't feel strongly. By making it the default, we're more likely to get feedback. On the other hand, it might be better to not change it now and then change it again soon if you think you're going to get around to styling it differently soon (with colors or underlining). Up to you.

File: lib/src/repo.rs [hunk 5]
```diff
@@ -1194,3 +1191,67 @@ mod dirty_cell {
...
+
+    /// This function returns the shortest length of a prefix of `key` that
+    /// corresponds to a trie that is either a) empty or b) contains only a
+    /// single element that matches `key` exactly.
+    ///
+    /// In the special case when there are keys in the trie for which our `key`
+    /// is an exact prefix, returns `key.len() + 1`. Conceptually, in order to
+    /// disambiguate, you need every letter of the key *and* the additional
+    /// fact that it's the entire key). This case is extremely unlikely for
+    /// hashes with 12+ hexadecimal characters.
```
[the remark this answers is not in the document]
Comment (ilyagr) [reply lib/src/repo.rs:1220]: Good point.

[the remark this answers is not in the document]
Comment (ilyagr) [reply lib/src/repo.rs]: Rewrote the comment, PTAL.

File: CHANGELOG.md (hunk 3 above)
Comment (ilyagr) [reply CHANGELOG.md]: 🤷

File: lib/src/repo.rs (hunk 4 above)
Comment (ilyagr) [reply lib/src/repo.rs:248]: Sorry about that. I'll try to make some future PR quick and efficient to review instead of planning on it being a huge pain.

**Comment by ilyagr** [comment]
> It's safe to merge two diverged repos by simply having one overwrite the other, as long as directories are merged in the trivial way (including both sets of files and subdirectories, not ignoring entries from one side). I think we'd lose that property if we started using IndexedLog.

This is a good point, by the way.

File: lib/src/repo.rs (hunk 4 above)
Comment (martinvonz) [reply lib/src/repo.rs:248]: No problem! It really wasn't a huge pain to review this PR :)

**Review by martinvonz — APPROVED** [review]
Thanks again for this feature!

**Comment by MForster** [comment]
One nice UI configuration I have seen with Mercurial (not sure if that's the default or if it's the Fig default): "highlight" the non-unique part of hashes in black-on-black. That reduces a lot of UI clutter, while at the same time still allowing copy-paste of the full hash.

Will this change allow this?

**Comment by martinvonz** [comment]
> "highlight" the non-unique part of hashes in black-on-black

Heh, that's a neat trick.

> Will this change allow this?

No, but I think Ilya might be working on some further changes to allow that. It required some work in the templater code, I think.

**Comment by ilyagr** [comment]
> Will this change allow this?

It's not yet possible to have different labels for different parts of the change/commit id. So, not quite yet.

I am looking into it, though I'm never sure how quickly I'll do anything. The templater code is confusing, I might have some questions. Or, I might through something inelegant that works and we can make it elegant later.

I personally would prefer grey-on-black, but once the labels work, you'd be able to configure any conceivable styling.

File: lib/src/repo.rs (hunk 4 above)
Comment (ilyagr) [reply lib/src/repo.rs:248]: I'm quite happy to hear that!

**Comment by MForster** [comment]
> I personally would prefer grey-on-black, but once the labels work, you'd be able to configure any conceivable styling.

Sure. The important part would be that it's configurable.

Exciting!
